pallets/pallet-asset-swap/src/xcm/trade.rs

```diff
@@ -36,71 +33,120 @@ mod xcm_fee_asset {
 	///
 	/// This trader is required in case there is no other mechanism to pay for
 	/// fees when transferring such an asset to this chain.
-	pub struct UsingComponentsForXcmFeeAsset<T: frame_system::Config, WeightToFee, Fungibles: Inspect<T::AccountId>>(
-		Weight,
-		Fungibles::Balance,
-		PhantomData<(T, WeightToFee)>,
-	);
+	pub struct UsingComponentsForXcmFeeAsset<T, WeightToFee>
+	where
+		T: Config,
+	{
+		remaining_weight: Weight,
+		remaining_fungible_balance: u128,
+		consumed_xcm_hash: Option<XcmHash>,
```

Comment:
Do you think tracking the xcm message is necessary? I would assume the messages are processed sequential. So there is no way two xcm message can collide

Reply:
Nope, that's not the correct assumption. If you see how the `impl_for_tuples` works, it sequentially calls `buy_weight` and then sequentially calls `refund_weight` and then `drop` in the end. So given the super-specific use cases in which these traders are to be used, we need to refund (and refund in drop) only if the very same asset has been withdrawn, which is why we need to have an `Option::Some` in there. Parity traders do something similar as well.

Reply:
Basically, we want to avoid the situation in which one trader is used to buy weight, and another one to refund it. We can control that by populating the state of the trader that was used to buy weight.

Reply:
Ah, okay. I looked into the PolkadotSdk and couldn't find something like that. That's why I was bringing it up here. Maybe I wasn't looking attentive enough

Reply:
You need to check the [runtimes repo](https://github.com/polkadot-fellows/runtimes)

Reply:
ah okay thank you